as.c

```diff
@@ -32,11 +33,338 @@
+}
+
+struct UrlComponents
+{
+    int port;
+    WCHAR host[URL_LEN];
+    bool https;
+};
+
+/**
+* Extracts protocol, port and hostname from URL
+*
+* @param url URL to parse, length must be less than URL_MAX
+*/
+void
+ParseUrl(const WCHAR *url, struct UrlComponents* comps)
+{
+    ZeroMemory(comps, sizeof(struct UrlComponents));
+
+    int domain_off = 0;
+    if (wcsbegins(url, L"http://")) {
+        comps->port = 80;
+        domain_off = 7;
+    } else if (wcsbegins(url, L"https://")) {
+        comps->port = 443;
+        comps->https = true;
+        domain_off = 8;
+    } else
+        comps->https = true;
+
+    WCHAR* strport = wcsstr(url + domain_off, L":");
+    if (strport) {
+        WCHAR* end;
+        wcsncpy(comps->host, url + domain_off, strport - url - domain_off);
+        comps->port = wcstol(strport + 1, &end, 10);
+    }
+    else {
+        if (comps->port == 0)
+            comps->port = 443;
+        wcscpy(comps->host, url + domain_off);
+    }
+
+    if (comps->host[wcslen(comps->host) - 1] == L'/')
+        comps->host[wcslen(comps->host) - 1] = L'\0';
+}
+
+/**
+ * Download profile content
+ *
+ * @param hWnd handle of window which initiated download
+ * @param hRequest WinInet request handle
+ * @param pbuf pointer to a buffer, will be allocated by this function. Caller must free it after use.
+ * @parap psize pointer to a profile size, assigned by this function
+ */
+BOOL
+DownloadProfileContent(HANDLE hWnd, HINTERNET hRequest, char** pbuf, size_t* psize)
+{
+    size_t pos = 0;
+    size_t size = READ_CHUNK_LEN;
+
+    *pbuf = calloc(1, size + 1);
+    char* buf = *pbuf;
+    if (buf == NULL) {
+        MessageBoxW(hWnd, L"Out of memory", _T(PACKAGE_NAME), MB_OK);
+        return FALSE;
+    }
+    while (true) {
+        DWORD bytesRead = 0;
+        if (!InternetReadFile(hRequest, buf + pos, READ_CHUNK_LEN, &bytesRead)) {
+            ShowWinInetError(hWnd);
+            return FALSE;
+        }
+        if (bytesRead == 0) {
+            size = pos;
+            break;
+        }
+
+        buf[pos + bytesRead] = '\0';
+
+        if (pos + bytesRead >= size) {
+            size += READ_CHUNK_LEN;
+            char* ptr = realloc(buf, size + 1);
+            if (!ptr) {
+                MessageBoxW(hWnd, L"Out of memory", _T(PACKAGE_NAME), MB_OK);
+                return FALSE;
+            }
+            buf = ptr;
+        }
+
+        pos += bytesRead;
+    }
+
+    *psize = size;
+
+    return TRUE;
+}
+
+/**
+ * Download profile from AS and save it to a special-named temp file
+ *
+ * @param hWnd handle of window which initiated download
+ * @param host AS hostname, entered by user, might contain protocol and port
+ * @param username UTF-8 encoded username used for HTTP basic auth
+ * @param password UTF-8 encoded password used for HTTP basic auth
+ * @param autologin should autologin profile be used
+ * @param out_path full path to where profile is downloaded. Value assigned by this function.
```

Comment:
Caller must ensure out_path is at least MAX_PATH. Better MAX_PATH + 1 as even microsoft is unsure whether +1 is needed :) I would prefer passing in the size of out_path from caller. Also see below.

Reply:
Added `out_path_size` parameter.